src/main/java/carpet/patches/EntityPlayerMPFake.java

```diff
@@ -11,45 +11,87 @@
-import net.minecraft.util.UserCache;
 import net.minecraft.util.registry.RegistryKey;
 import net.minecraft.world.GameMode;
 import net.minecraft.world.World;
-import net.minecraft.world.dimension.DimensionType;
 import carpet.fakes.ServerPlayerEntityInterface;
 import carpet.utils.Messenger;
 
+import java.util.Objects;
+
 @SuppressWarnings("EntityConstructor")
 public class EntityPlayerMPFake extends ServerPlayerEntity
 {
     public Runnable fixStartingPosition = () -> {};
     public boolean isAShadow;
 
-    public static EntityPlayerMPFake createFake(String username, MinecraftServer server, double d0, double d1, double d2, double yaw, double pitch, RegistryKey<World> dimensionId, GameMode gamemode)
-    {
-        //prolly half of that crap is not necessary, but it works
-        ServerWorld worldIn = server.getWorld(dimensionId);
-        ServerPlayerInteractionManager interactionManagerIn = new ServerPlayerInteractionManager(worldIn);
-        UserCache.setUseRemote(false);
-        GameProfile gameprofile;
-        try {
-            gameprofile = server.getUserCache().findByName(username);
+    public static EntityPlayerMPFake createFake(String playerName, MinecraftServer server, double d0, double d1, double d2, double yaw, double pitch, RegistryKey<World> dimensionId, GameMode gamemode, boolean isSpawnerPrivileged, boolean fallBackToSpawningOfflinePlayer
+    ) throws FakePlayerSpawnException {
+        Objects.requireNonNull(playerName);
+        Objects.requireNonNull(server);
+        Objects.requireNonNull(dimensionId);
+        Objects.requireNonNull(gamemode);
+
+        PlayerManager playerManager = server.getPlayerManager();
+
+        // if this player already logged in
+        if (playerManager.getPlayer(playerName) != null)
+        {
+            throw new FakePlayerSpawnException("r Player ", "rb " + playerName, "r  is already logged on");
+        }
+
+        // check player name length
+        // player with a name longer than 16 bytes will kick out all players
+        if (playerName.length() > 16)
+            throw new FakePlayerSpawnException("rb Player name: " + playerName + " is too long");
+
+        // get player profile from auth server (or generate offline-mode profile if the server is in offline mode)
+        GameProfile gameprofile = server.getUserCache().findByName(playerName);
+
+        // if the player with this playerName does not exist
+        if (gameprofile == null)
+        {
+            if (!fallBackToSpawningOfflinePlayer) {
+                throw new FakePlayerSpawnException("r Player " + playerName + " is either banned by Mojang, or " +
+                        "auth servers are down. Banned players can only be summoned in Singleplayer " +
+                        "and in servers in off-line mode.");
+            } else {
+                // spawn offline player
+                gameprofile = new GameProfile(PlayerEntity.getOfflinePlayerUuid(playerName), playerName);
```

Comment:
isn't that THE only line that really needed to be added to have the mvp in this situation?

Reply:
No, it isn't.
In file `PlayerCommand.java`, method `cantSpawn` checks the specified fake player name:

```java
    private static boolean cantSpawn(CommandContext<ServerCommandSource> context)
    {
        String playerName = StringArgumentType.getString(context, "player");
        MinecraftServer server = context.getSource().getMinecraftServer();
        PlayerManager manager = server.getPlayerManager();
        PlayerEntity player = manager.getPlayer(playerName);
        if (player != null)
        {
            Messenger.m(context.getSource(), "r Player ", "rb " + playerName, "r  is already logged on");
            return true;
        }
        GameProfile profile = server.getUserCache().findByName(playerName);
        if (profile == null)
        {
            Messenger.m(context.getSource(), "r Player "+playerName+" is either banned by Mojang, or auth servers are down. " +
                    "Banned players can only be summoned in Singleplayer and in servers in off-line mode.");
            return true;
        }
// ...
```

And if the online-mode player with specified name does not exist, this method will return `true`, which makes `spawn` method in `PlayerCommand.java` returns immediately:

```java
    private static int spawn(CommandContext<ServerCommandSource> context) throws CommandSyntaxException
    {
        if (cantSpawn(context)) return 0;
// ...
```

If only the line you mentioned was added, that line will never be reached.

Reply:
well, I added the other comment above later
Allow it there (by the new rule): https://github.com/gnembon/fabric-carpet/pull/872/files#r637356921
and make it happen here, right?

Reply:
Yeah I re-implemented this feature in that way.